### PR TITLE
【BIT】paddle2torch转换 paddle.vision.ops.psroi_pool paddle.vision.ops.roi_align paddle.vision.ops.roi_pool paddle.tensor.gather paddle.tensor.gather_nd

### PR DESCRIPTION
**Pull request by cszdrg**
## paddle.vision.ops.psroi_pool paddle.vision.ops.roi_align paddle.vision.ops.roi_pool
https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/guides/model_convert/convert_from_pytorch/api_difference_third_party/torchvision/torchvision.ops.ps_roi_pool.html
https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/guides/model_convert/convert_from_pytorch/api_difference_third_party/torchvision/torchvision.ops.roi_align.html
https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/guides/model_convert/convert_from_pytorch/api_difference_third_party/torchvision/torchvision.ops.roi_pool.html
三个api处理思路大致相同，args基本可以对应，但是torch中的boxes被对应到paddle中的boxes和boxes_num两个参数，所以需要进行处理

a）经发现 api_config中的boxes_num参数生成存在问题，进行调整输出合适的tensor
b）创建rule处理boxes和boxes_num两个参数
c）存在精度问题，放入report中
d) roi_pool和psroi_pool参数一致 采用相同rule

## paddle.tensor.gather paddle.tensor.gather_nd
https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/api/paddle/gather_cn.html
https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/api/paddle/gather_nd_cn.html
这个两个api在torch中没有对应api
a)采用torch.narrow将对应维度抽出
b)使用torch.stack进行拼接

### CODE REVIEW

File: tester/paddle_to_torch/rules.py [hunk 1]
```diff
@@ -314,12 +314,87 @@ def apply(self, paddle_api: str) -> ConvertResult:
 
 # p
 
-
+class psroi_poolRule(BaseRule):
```
Comment (cangtianhuang) [inline tester/paddle_to_torch/rules.py:317]: 类名应采用首字母大写的驼峰命名法

File: tester/paddle_to_torch/rules.py [hunk 2]
```diff
@@ -314,12 +314,87 @@ def apply(self, paddle_api: str) -> ConvertResult:
...
+result = torchvision.ops.ps_roi_pool( **_kwargs, boxes = ans)
+"""
+        code = impl.splitlines()
+        return ConvertResult.success(paddle_api, code, "result")
 # q
 
 
 # r
-
+class roi_alignRule(BaseRule):
```
Comment (cangtianhuang) [inline tester/paddle_to_torch/rules.py:345]: 同上

File: tester/paddle_to_torch/rules.py [hunk 3]
```diff
@@ -314,12 +314,87 @@ def apply(self, paddle_api: str) -> ConvertResult:
...
+    else:
+        _kwargs[torch_param] = 1.0
+boxes = locals().get('boxes')
+boxnum = locals().get('boxes_num')
+ans = []
+end = 0
+for i in range(boxnum.shape[0]):
+    begin = end
+    end = end + int(boxnum[i])
+    ans.append(boxes[begin:end,])
```
Comment (cangtianhuang) [inline tester/paddle_to_torch/rules.py:393]: 感觉这三个rule的逻辑是可复用的，能写成一个Rule吗？通过api name区分？

File: tester/paddle_to_torch/rules.py [hunk 4]
```diff
@@ -314,12 +314,87 @@ def apply(self, paddle_api: str) -> ConvertResult:
...
+result = torchvision.ops.roi_pool(input = _kwargs['input'],boxes = ans, output_size = _kwargs['output_size'], spatial_scale = _kwargs['spatial_scale'])
```
Comment (cangtianhuang) [inline tester/paddle_to_torch/rules.py:394]:
```suggestion
result = torchvision.ops.roi_pool(boxes = ans, **_kwargs)
```

**Review by cangtianhuang — CHANGES_REQUESTED** [review]

File: tester/paddle_to_torch/rules.py (hunk 3 above)
Comment (cszdrg) [reply tester/paddle_to_torch/rules.py]: 参数有些不一样 还是区分开吧

Comment (cszdrg) [reply tester/paddle_to_torch/rules.py]: roi_pool和psroi_pool参数一致，就放到一起了
roi_align参数不一致 还涉及到默认值的问题，就单独领出来了

File: tester/paddle_to_torch/mapping.json [hunk 5]
```diff
@@ -168,8 +174,19 @@
...
+    },
     "q": "API started with 'q' should be placed here",
     "r": "API started with 'r' should be placed here",
+    "paddle.vision.ops.roi_align":{
+        "Rule": "Roi_aignRule"
+    },
+    "paddle.vision.ops.roi_pool": {
+        "Rule": "Roi_poolRule",
+        "torch_api":"torchvision.ops.roi_pool"
+    },
```
Comment (cangtianhuang) [inline tester/paddle_to_torch/mapping.json:189]:
```suggestion
    "paddle.vision.ops.roi_pool": {
        "torch_api":"torchvision.ops.roi_pool",
        "Rule": "Roi_poolRule"
    },
```

File: tester/paddle_to_torch/rules.py [hunk 6]
```diff
@@ -96,6 +96,8 @@ def read_mapping(self, mapping: Dict):
         """
         self.mapping: Dict = mapping
         if "Rule" in mapping:
+            if "torch_api" in mapping:
+                self.torch_api: str = mapping["torch_api"]
             return
```
Comment (cangtianhuang) [inline tester/paddle_to_torch/rules.py:101]: 这部分可能之后还要改一改，允许rule的同时添加参数和映射表。现在先这样。

File: tester/paddle_to_torch/rules.py [hunk 7]
```diff
@@ -302,7 +304,47 @@ def apply(self, paddle_api: str) -> ConvertResult:
...
+class GatherRule(BaseRule):
+    def apply(self, paddle_api: str) -> ConvertResult:
+        #抽取对应维度的tensor直接进行stack操作
+        impl = """
+x = locals().get('x')
+index = locals().get('index')
+if 'axis' in locals():
+    axis = locals().get('axis')
+else:
+    axis = 0
```
Comment (cangtianhuang) [inline tester/paddle_to_torch/rules.py:316]:
```suggestion
axis = locals().get('axis', 0)
```

File: tester/paddle_to_torch/rules.py [hunk 8]
```diff
@@ -302,7 +304,47 @@ def apply(self, paddle_api: str) -> ConvertResult:
...
+                temp = torch.squeeze(temp, 0)
+            return temp
+        ans = []
+        for i in index:
+            ans.append(self.func(x, i))
+        return torch.stack(ans, 0)
+f = _func()
+x = locals().get('x')
+index = locals().get('index')
+result = f.func(x,index)
```
Comment (cangtianhuang) [inline tester/paddle_to_torch/rules.py:344]: 可以直接def func吧，不需要定义类了

**Review by cangtianhuang — CHANGES_REQUESTED** [review]
tester/api_config/api_config_support2torch_8.txt 更名为 tester/api_config/api_config_support2torch_4.txt，每人认领两个文件。不需要对应 merged 😊

File: tester/paddle_to_torch/rules.py (hunk 8 above)
Comment (cszdrg) [reply tester/paddle_to_torch/rules.py:344]: 不可以 开始就是定义方法的 作用域会出问题 可能是奇怪的bug

File: tester/paddle_to_torch/rules.py (hunk 7 above)
Comment (cszdrg) [reply tester/paddle_to_torch/rules.py:316]: 为什么要加到locals里呢？

File: tester/paddle_to_torch/mapping.json (hunk 5 above)
Comment (cszdrg) [reply tester/paddle_to_torch/mapping.json:189]: 为啥要调换？这样不行吗？

File: tester/paddle_to_torch/rules.py (hunk 6 above)
Comment (cszdrg) [reply tester/paddle_to_torch/rules.py:101]: ok

File: tester/paddle_to_torch/rules.py [hunk 9]
```diff
@@ -302,7 +304,47 @@ def apply(self, paddle_api: str) -> ConvertResult:
...
+index = locals().get('index')
+if 'axis' in locals():
+    axis = locals().get('axis')
+else:
+    axis = 0
+ans = []
+for i in index:
+    ans.append(torch.narrow(x, axis, i.reshape([]), 1))
+result = torch.stack(ans,axis)
+result = torch.squeeze(result)
```
Comment (Cutelemon6) [inline tester/paddle_to_torch/rules.py:321]: paddle.gather 和 torch.narrow 的区别在哪里，从文档看是等价的

**Comment by cszdrg** [comment]
> tester/api_config/api_config_support2torch_8.txt 更名为 tester/api_config/api_config_support2torch_4.txt，每人认领两个文件。不需要对应 merged 😊

是的 现在刚开始 我怕大家命名冲突 等下几个pr会找个没人用的命名改回去

File: tester/paddle_to_torch/rules.py (hunk 7 above)
Comment (cszdrg) [reply tester/paddle_to_torch/rules.py:316]: 后面还会用到这个locals()里的参数吗？如果没有用到 这样子运行速度会快一点吧

Comment (cangtianhuang) [reply tester/paddle_to_torch/rules.py:316]: 这不是加到locals里，这是为get方法添加默认值（locals没有就是后面的值），和你的代码逻辑是一样的。

File: tester/paddle_to_torch/rules.py (hunk 8 above)
Comment (cangtianhuang) [reply tester/paddle_to_torch/rules.py:344]: ？？好家伙，好的，可以把这类问题记录到QA里，避免其他同学踩坑。不知道前面添加空行可不可以。

**Comment by cangtianhuang** [comment]
> tester/api_config/api_config_support2torch_8.txt 更名为 tester/api_config/api_config_support2torch_4.txt，每人认领两个文件。不需要对应 merged 😊

现在我1xym2mzj3，wyh还没领，你就认领4呗～

File: tester/paddle_to_torch/rules.py (hunk 9 above)
Comment (cszdrg) [reply tester/paddle_to_torch/rules.py:321]: narrow只能选取连续区间的维度，gather随心所欲，还可以重复选取

**Comment by cszdrg** [comment]
> > tester/api_config/api_config_support2torch_8.txt 更名为 tester/api_config/api_config_support2torch_4.txt，每人认领两个文件。不需要对应 merged 😊
> 
> 现在我1xym2mzj3，wyh还没领，你就认领4呗～

行

File: report/fresh_report/accuracy/test_log_roi.txt [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+paddle.vision.ops.roi_align(Tensor([1, 256, 32, 32],"float32"), Tensor([3, 4],"float32"), boxes_num=Tensor([1],"int32"), output_size=3, )
+paddle.vision.ops.roi_align(Tensor([1, 256, 32, 32],"float32"), Tensor([3, 4],"float32"), boxes_num=Tensor([1],"int32"), output_size=tuple(3,4,), )
+
+
+grep: warning: GREP_OPTIONS is deprecated; please use an alias or script
+2025-04-23 05:33:10.429133 test begin: paddle.vision.ops.psroi_pool(Tensor([2, 128, 32, 32],"float32"), Tensor([4, 4],"float32"), Tensor([2],"int32"), 8, 1.1, )
+
+W0423 05:33:13.055356 16186 gpu_resources.cc:119] Please NOTE: device: 0, GPU Compute Capability: 6.1, Driver API Version: 12.0, Runtime API Version: 12.0
+W0423 05:33:13.056680 16186 gpu_resources.cc:164] device: 0, cuDNN Version: 8.8.
+One of the differentiated Tensors does not require grad
```
[the remark this answers is not in the document]
Comment (cszdrg) [reply report/fresh_report/accuracy/test_log_roi.txt:16]: 这确实有点奇怪 api问题或者引擎有问题？ 我有空去看看

**Review by Cutelemon6 — APPROVED** [review]
LGTM

**Review by cangtianhuang — APPROVED** [review]
LGTM
fresh_report 里的精度出错问题皇润同学可以进一步定位一下，以确定是引擎问题还是 api 精度问题🫡

**Comment by cszdrg** [comment]
> LGTM fresh_report 里的精度出错问题皇润同学可以进一步定位一下，以确定是引擎问题还是 api 精度问题🫡

这个报错是指torch中的某个input不进行梯度计算 没有问题
@cangtianhuang  @Cutelemon6

**Comment by Cutelemon6** [comment]
需要检查一下引擎的相关代码，我认为 paddle2torch 应该保证要么都计算梯度要么都不计算

File: tester/api_config/test_log/checkpoint.txt [hunk 11]
```diff
@@ -1,881 +1,37 @@
-paddle.all(Tensor([1, 2281701379],"bool"), )
-paddle.amax(Tensor([2, 1073741824, 2],"int32"), tuple(0,1,), False, )
-paddle.amin(Tensor([3, 2, 95070891, 4],"float32"), axis=2, keepdim=True, )
-paddle.mean(Tensor([1, 3, 5, 57266231, 5],"float16"), )
```
Comment (wanghuancoder) [inline tester/api_config/test_log/checkpoint.txt:4]: 把这里面的内容删掉

**Review by wanghuancoder — APPROVED** [review]
LGTM
